alumet/src/pipeline/control.rs

```diff
@@ -0,0 +1,213 @@
+use super::builder::elements::{
+    AutonomousSourceBuilder, ManagedSourceBuilder, ManagedSourceRegistration, SendSourceBuilder,
+};
+use super::elements::{output, source, transform};
+use super::{builder, trigger, PluginName, Source};
+use tokio::runtime;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::task::JoinHandle;
+use tokio_util::sync::CancellationToken;
+
+#[derive(Clone)]
+pub struct AnonymousControlHandle {
+    tx: Sender<ControlMessage>,
+    shutdown: CancellationToken,
+}
+
+#[derive(Clone)]
+pub struct ScopedControlHandle {
+    inner: AnonymousControlHandle,
+    plugin: PluginName,
+}
+
+pub enum ControlMessage {
+    Source(source::ControlMessage),
+    Transform(transform::ControlMessage),
+    Output(output::ControlMessage),
+}
+
+pub(crate) struct PipelineControl {
+    sources: source::SourceControl,
+    transforms: transform::TransformControl,
+    outputs: output::OutputControl,
+}
+
+pub enum ControlError {
+    ChannelFull(ControlMessage),
+    Shutdown,
+}
+
+impl AnonymousControlHandle {
+    pub async fn send(&mut self, message: ControlMessage) -> Result<(), ControlError> {
+        self.tx.send(message).await.map_err(|_| ControlError::Shutdown)
+    }
+
+    pub fn try_send(&mut self, message: ControlMessage) -> Result<(), ControlError> {
+        match self.tx.try_send(message) {
+            Ok(_) => Ok(()),
+            Err(mpsc::error::TrySendError::Full(m)) => Err(ControlError::ChannelFull(m)),
+            Err(mpsc::error::TrySendError::Closed(_)) => Err(ControlError::Shutdown),
+        }
+    }
+
+    pub fn shutdown(&mut self) {
+        self.shutdown.cancel()
+    }
+
+    pub fn scoped(&self, plugin: PluginName) -> ScopedControlHandle {
+        ScopedControlHandle {
+            inner: self.clone(),
+            plugin,
+        }
+    }
+}
+
+impl ScopedControlHandle {
+    pub fn change_plugin_scope(&self, plugin: PluginName) -> ScopedControlHandle {
+        ScopedControlHandle {
+            inner: self.inner.clone(),
+            plugin,
+        }
+    }
```

Comment:
the scope is the plugin, it comes from the name of the `ScopedControlHandle` struct. It means that everything that this handle does has the plugin as its scope, for instance sources will be "tagged" with the plugin's name.

If you find it confusing, maybe the naming could be better?
- `PluginControlHandle` is bad because it makes the reader think that the handle controls a plugin, which is wrong
- `NamedControlHandle` ?
- `ControlHandleForPlugin` ?

Reply:
Nah it's ok, with your comment I understand why you did it